File: lib/features/damage_calculator/domain/damage_calculator_session.dart
import 'package:clashkingapp/features/damage_calculator/data/damage_catalog.dart';
import 'package:clashkingapp/features/damage_calculator/domain/damage_calculator_engine.dart';

class DamageAccountPreset {
  const DamageAccountPreset({
    required this.tag,
    required this.name,
    required this.townHall,
    this.ownedLevels = const {},
  });

  final String tag;
  final String name;
  final int townHall;
  final Map<DamageSourceKind, int> ownedLevels;
}

class SelectedBuilding {
  const SelectedBuilding({required this.buildingId, required this.level});

  final String buildingId;
  final int level;

  SelectedBuilding copyWith({int? level}) =>
      SelectedBuilding(buildingId: buildingId, level: level ?? this.level);
}

class SelectedDamageSource {
  const SelectedDamageSource({
    required this.kind,
    required this.level,
    required this.count,
  });

  final DamageSourceKind kind;
  final int level;
  final int count;

  SelectedDamageSource copyWith({int? level, int? count}) =>
      SelectedDamageSource(
        kind: kind,
        level: level ?? this.level,
        count: count ?? this.count,
      );
}

class DamageCalculatorSession {
  DamageCalculatorSession(this.catalog, {int? townHall})
    : townHall = (townHall ?? catalog.maxTownHall).clamp(
        1,
        catalog.maxTownHall,
      ) {
    _seedDefaults();
  }

  final DamageCatalog catalog;
  int townHall;
  int spellCapacity = 11;
  String? selectedAccountTag;
  final List<SelectedBuilding> targets = [];
  final Map<DamageSourceKind, SelectedDamageSource> sources = {};

  List<BuildingDefinition> get availableBuildings =>
      catalog.buildingsForTownHall(townHall);

  List<DamageSourceDefinition> get availableSources => catalog.sources
      .where((source) => source.levelsForTownHall(townHall).isNotEmpty)
      .toList(growable: false);

  void setTownHall(int value) {
    townHall = value.clamp(1, catalog.maxTownHall);
    _repairTargets();
    _repairSources();
  }

  bool addTarget(String buildingId) {
    if (targets.any((target) => target.buildingId == buildingId)) return false;
    final building = _building(buildingId);
    if (building == null) return false;
    final levels = building.levelsForTownHall(townHall);
    if (levels.isEmpty) return false;
    targets.add(
      SelectedBuilding(buildingId: buildingId, level: levels.last.level),
    );
    return true;
  }

  void removeTarget(String buildingId) {
    targets.removeWhere((target) => target.buildingId == buildingId);
  }

  void setTargetLevel(String buildingId, int level) {
    final building = _building(buildingId);
    if (building == null) return;
    final valid = building.levelsForTownHall(townHall);
    if (!valid.any((candidate) => candidate.level == level)) return;
    final index = targets.indexWhere(
      (target) => target.buildingId == buildingId,
    );
    if (index >= 0) targets[index] = targets[index].copyWith(level: level);
  }

  void setSourceLevel(DamageSourceKind kind, int level) {
    final definition = catalog.source(kind);
    if (definition == null) return;
    final valid = definition.levelsForTownHall(townHall);
    if (!valid.any((candidate) => candidate.level == level)) return;
    final current = sources[kind];
    sources[kind] = SelectedDamageSource(
      kind: kind,
      level: level,
      count: current?.count ?? 0,
    );
  }

  void setSourceCount(DamageSourceKind kind, int count) {
    final current = sources[kind];
    if (current == null) return;
    sources[kind] = current.copyWith(count: count.clamp(0, 99));
  }

  void setSpellCapacity(int value) {
    spellCapacity = value.clamp(1, 20);
  }

  void applyPreset(DamageAccountPreset preset) {
    selectedAccountTag = preset.tag;
    setTownHall(preset.townHall);
    for (final entry in preset.ownedLevels.entries) {
      final definition = catalog.source(entry.key);
      if (definition == null) continue;
      final valid = definition.levelsForTownHall(townHall);
      if (valid.isEmpty) continue;
      final chosen = valid.lastWhere(
        (level) => level.level <= entry.value,
        orElse: () => valid.first,
      );
      setSourceLevel(entry.key, chosen.level);
    }
  }

  List<DamageTarget> resolvedTargets() {
    final resolved = <DamageTarget>[];
    for (final selected in targets) {
      final building = _building(selected.buildingId);
      final level = building?.level(selected.level);
      if (building != null && level != null) {
        resolved.add(DamageTarget(building: building, level: level));
      }
    }
    return resolved;
  }

  List<DamageStackEntry> resolvedStack() {
    final resolved = <DamageStackEntry>[];
    for (final selected in sources.values) {
      final source = catalog.source(selected.kind);
      final level = source?.level(selected.level);
      if (source != null && level != null && selected.count > 0) {
        resolved.add(
          DamageStackEntry(source: source, level: level, count: selected.count),
        );
      }
    }
    return resolved;
  }

  void _seedDefaults() {
    _repairSources();
    final townHallBuilding = catalog.buildings
        .where((building) => building.name == 'Town Hall')
        .firstOrNull;
    if (townHallBuilding != null) addTarget(townHallBuilding.id);
  }

  void _repairTargets() {
    for (var index = targets.length - 1; index >= 0; index--) {
      final selected = targets[index];
      final building = _building(selected.buildingId);
      final valid = building?.levelsForTownHall(townHall) ?? const [];
      if (valid.isEmpty) {
        targets.removeAt(index);
        continue;
      }
      if (!valid.any((level) => level.level == selected.level)) {
        targets[index] = selected.copyWith(level: valid.last.level);
      }
    }
  }

  void _repairSources() {
    final validKinds = <DamageSourceKind>{};
    for (final source in catalog.sources) {
      final valid = source.levelsForTownHall(townHall);
      if (valid.isEmpty) continue;
      validKinds.add(source.kind);
      final current = sources[source.kind];
      final currentValid =
          current != null && valid.any((level) => level.level == current.level);
      sources[source.kind] = SelectedDamageSource(
        kind: source.kind,
        level: currentValid ? current.level : valid.last.level,
        count: current?.count ?? 0,
      );
    }
    sources.removeWhere((kind, _) => !validKinds.contains(kind));
  }

  BuildingDefinition? _building(String id) {
    for (final building in catalog.buildings) {
      if (building.id == id) return building;
    }
    return null;
  }
}

extension<T> on Iterable<T> {
  T? get firstOrNull => isEmpty ? null : first;
}

File: expo/src/features/damage-calculator/domain/damage-calculator-session.ts
import { DamageCatalog } from '../data/damage-catalog';
import {
  type BuildingDefinition,
  type DamageStackEntry,
  DamageSourceKind,
  DamageTarget,
} from './damage-calculator-engine';

export interface DamageAccountPreset {
  readonly tag: string;
  readonly name: string;
  readonly townHall: number;
  readonly league?: string;
  readonly ownedLevels?: ReadonlyMap<DamageSourceKind, number>;
}

export interface SelectedBuilding {
  readonly buildingId: string;
  readonly level: number;
}

export interface SelectedDamageSource {
  readonly kind: DamageSourceKind;
  readonly level: number;
  readonly count: number;
}

export class DamageCalculatorSession {
  readonly catalog: DamageCatalog;
  townHall: number;
  spellCapacity = 11;
  selectedAccountTag?: string;
  readonly targets: SelectedBuilding[] = [];
  readonly sources = new Map<DamageSourceKind, SelectedDamageSource>();

  constructor(catalog: DamageCatalog, options: { townHall?: number } = {}) {
    this.catalog = catalog;
    this.townHall = clamp(options.townHall ?? catalog.maxTownHall, 1, catalog.maxTownHall);
    this.repairSources();
  }

  get availableBuildings(): readonly BuildingDefinition[] {
    return this.catalog.buildingsForTownHall(this.townHall);
  }

  get availableSources() {
    return this.catalog.sources.filter(
      (source) => source.levelsForTownHall(this.townHall).length > 0,
    );
  }

  setTownHall(value: number): void {
    this.townHall = clamp(value, 1, this.catalog.maxTownHall);
    this.repairTargets();
    this.repairSources();
  }

  addTarget(buildingId: string): boolean {
    if (this.targets.some((target) => target.buildingId === buildingId)) return false;
    const building = this.availableBuilding(buildingId);
    if (!building) return false;
    const levels = building.levelsForTownHall(this.townHall);
    const last = levels.at(-1);
    if (!last) return false;
    this.targets.push({ buildingId, level: last.level });
    return true;
  }

  removeTarget(buildingId: string): void {
    const index = this.targets.findIndex((target) => target.buildingId === buildingId);
    if (index >= 0) this.targets.splice(index, 1);
  }

  setTargetLevel(buildingId: string, level: number): void {
    const building = this.availableBuilding(buildingId);
    if (!building) return;
    const valid = building.levelsForTownHall(this.townHall);
    if (!valid.some((candidate) => candidate.level === level)) return;
    const index = this.targets.findIndex((target) => target.buildingId === buildingId);
    if (index >= 0) this.targets[index] = { buildingId, level };
  }

  setSourceLevel(kind: DamageSourceKind, level: number): void {
    const definition = this.catalog.source(kind);
    if (!definition) return;
    const valid = definition.levelsForTownHall(this.townHall);
    if (!valid.some((candidate) => candidate.level === level)) return;
    this.sources.set(kind, {
      kind,
      level,
      count: this.sources.get(kind)?.count ?? 0,
    });
  }

  setSourceCount(kind: DamageSourceKind, count: number): void {
    const current = this.sources.get(kind);
    if (!current) return;
    this.sources.set(kind, { ...current, count: clamp(count, 0, 99) });
  }

  setSpellCapacity(value: number): void {
    this.spellCapacity = clamp(value, 1, 20);
  }

  applyPreset(preset: DamageAccountPreset): void {
    this.selectedAccountTag = preset.tag;
    this.setTownHall(preset.townHall);
    for (const [kind, ownedLevel] of preset.ownedLevels ?? []) {
      const definition = this.catalog.source(kind);
      if (!definition) continue;
      const valid = definition.levelsForTownHall(this.townHall);
      if (valid.length === 0) continue;
      const chosen = [...valid].reverse().find((level) => level.level <= ownedLevel) ?? valid[0]!;
      this.setSourceLevel(kind, chosen.level);
    }
  }

  resolvedTargets(): readonly DamageTarget[] {
    const resolved: DamageTarget[] = [];
    for (const selected of this.targets) {
      const building = this.availableBuilding(selected.buildingId);
      const level = building?.level(selected.level);
      if (building && level) resolved.push(new DamageTarget({ building, level }));
    }
    return resolved;
  }

  resolvedStack(): readonly DamageStackEntry[] {
    const resolved: DamageStackEntry[] = [];
    for (const selected of this.sources.values()) {
      const source = this.catalog.source(selected.kind);
      const level = source?.level(selected.level);
      if (source && level && selected.count > 0) {
        resolved.push({ source, level, count: selected.count });
      }
    }
    return resolved;
  }

  private repairTargets(): void {
    for (let index = this.targets.length - 1; index >= 0; index -= 1) {
      const selected = this.targets[index]!;
      const building = this.availableBuilding(selected.buildingId);
      const valid = building?.levelsForTownHall(this.townHall) ?? [];
      const last = valid.at(-1);
      if (!last) {
        this.targets.splice(index, 1);
      } else if (!valid.some((level) => level.level === selected.level)) {
        this.targets[index] = { ...selected, level: last.level };
      }
    }
  }

  private repairSources(): void {
    const validKinds = new Set<DamageSourceKind>();
    for (const source of this.catalog.sources) {
      const valid = source.levelsForTownHall(this.townHall);
      const last = valid.at(-1);
      if (!last) continue;
      validKinds.add(source.kind);
      const current = this.sources.get(source.kind);
      const currentValid =
        current !== undefined && valid.some((level) => level.level === current.level);
      this.sources.set(source.kind, {
        kind: source.kind,
        level: currentValid ? current.level : last.level,
        count: current?.count ?? 0,
      });
    }
    for (const kind of this.sources.keys()) {
      if (!validKinds.has(kind)) this.sources.delete(kind);
    }
  }

  private availableBuilding(id: string): BuildingDefinition | undefined {
    return this.availableBuildings.find((building) => building.id === id);
  }
}

function clamp(value: number, minimum: number, maximum: number): number {
  return Math.min(maximum, Math.max(minimum, value));
}
